javascript/ql/src/semmle/javascript/Expr.qll

```diff
@@ -161,6 +161,58 @@ class Expr extends @expr, ExprOrStmt, ExprOrType, AST::ValueNode {
    * file that was extracted without type information.
    */
   Type getType() { ast_node_type(this, result) }
+
+  /**
+   * Holds if the syntactic context that the expression appears in relies on the expression
+   * being non-null/non-undefined.
+   */
+  predicate inNullSensitiveContext() {
+    exists(ExprOrStmt ctx |
+      // bases cases
+      this = ctx.(PropAccess).getBase()
+      or
+      this = ctx.(IndexExpr).getPropertyNameExpr()
+      or
+      this = ctx.(InvokeExpr).getCallee()
+      or
+      this = ctx.(BinaryExpr).getAnOperand() and
+      not ctx instanceof LogicalBinaryExpr and // x LOGOP y is fine because of implicit conversion
+      not ctx instanceof EqualityTest and // x EQOP y is fine because of implicit conversion and lack thereof
+      not ctx.(BitOrExpr).getAnOperand().(NumberLiteral).getIntValue() = 0 and // x | 0 is fine because it's used to convert to numbers
+      not ctx.(RShiftExpr).getRightOperand().(NumberLiteral).getIntValue() = 0 and // x >> 0 is fine because it's used to convert to numbers
+      not ctx.(URShiftExpr).getRightOperand().(NumberLiteral).getIntValue() = 0 // x >> 0 is fine because it's used to convert to numbers
+      or
+      this = ctx.(UnaryExpr).getOperand() and
+      not ctx instanceof LogNotExpr and // !x is fine because of implicit conversion
+      not ctx instanceof PlusExpr // +x is fine because of implicit conversion
+      or
+      this = ctx.(UpdateExpr).getOperand()
+      or
+      this = ctx.(CompoundAssignExpr).getLhs()
+      or
+      this = ctx.(CompoundAssignExpr).getRhs()
+      or
+      this = ctx.(AssignExpr).getRhs() and
+      ctx.(AssignExpr).getLhs() instanceof DestructuringPattern
+      or
+      this = ctx.(SpreadElement).getOperand()
+      or
+      this = ctx.(ForOfStmt).getIterationDomain()
+      or
+      // recursive cases
+      this = ctx.(ParExpr).getExpression() and
+      ctx.(ParExpr).inNullSensitiveContext()
```

Comment:
```suggestion
      ctx.inNullSensitiveContext()
```

Reply:
Similarly below.

Reply:
I would love to do this except the QL compiler complaints that `ctx` is not a `ParExpr` if I remove the second cast. :(

Reply:
Ah, you've generalised `ctx` to be an `ExprOrStmt` to accommodate the `ForOfStmt` case; that's why you need a cast.